Simplify c10 dispatcher

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29682 backend fallback test
* #29681 better boxing
* #29337 Hide the OperatorKernel* argument from the stack based kernel API
* #29330 OperatorHandle::callBoxed/callUnboxed
* #29214 Test cases for backend fallback kernels
* #29201 Add a OperatorHandle argument to boxed kernels
* #28494 Implement backend level fallback for c10
* **#28314 Simplify c10 dispatcher**

Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

smessmer committed:
Simplify c10 dispatcher

Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -0,0 +1,85 @@
+
+  template<class... Args>
+  c10::optional<TensorTypeId> getDispatchKeyUnboxed(const Args&... args) const {
+    AT_ASSERT(sizeof...(args) == num_args_, "Wrong number of arguments. Schema says ", num_args_, " but was called with ", sizeof...(args));
```
this check is in the hot inlined path, isn't it? did you check that it doesn't blow up the codesize? also, do we need this check at all? if we assume that unboxed version is called by codegen only, it's probably safe to assume that it's ok. Or we can make it a DCHECK or something

smessmer replied:
I'll remove the check.

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  c10::optional<TensorTypeId> getDispatchKeyBoxed(const Stack* stack) const {
+    if (C10_UNLIKELY(!is_valid_)) {
```
I don't want to address it in this PR because that would be a more intrusive change and I'm not sure I find time for it later, but I think it's a good idea.

smessmer replied:
fixed

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -0,0 +1,85 @@
+ */
+struct DispatchKeyExtractor final {
+public:
+  static DispatchKeyExtractor make(const FunctionSchema& schema) {
```
it's easier to write this way. As a constructor, I wouldn't be able to use constructor initializer lists because there's a whole lot of code that needs to run before I know what num_args or is_valid are supposed to be.

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -121,157 +77,71 @@ class DispatchTable final {
-      // variable
-      return c10::legacyExtractTypeId(ts);
+  std::string listAllDispatchKeys() const {
+    if (kernels_.size() == 0) {
```
fixed

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -121,157 +77,71 @@ class DispatchTable final {
   void setCatchallKernel(const KernelFunction& kernel) {
-    if (catchall_kernel_.has_value()) {
-      TORCH_WARN("Registered a catch-all kernel for operator ", operator_name_," that overwrote a previously registered catch-all kernel for the same operator.");
+    if (catchallKernel_.has_value()) {
```
it needs to have a lower priority than the backend level fallbacks (i.e. lazy), so a simple "always set" typeid won't work.

ezyang replied:
No, I think this suggestion would work. Basically, the idea is that every tensor has GenericTensorTypeId in the tensor type id set (so, a normal looking tensor, would be something like GenericTensorTypeId, CPUTensorId, VariableTensorId). And you just place GenericTensorTypeId all the way at the end of the dispatch chain.

In any case, this shouldn't happen in this PR.

smessmer replied:
but how would this fallback to the GenericTensorTypeId? We only look at the highest id, we don't "fall back" to lower ids if the highest one doesn't match.

ezyang replied:
I guess, maybe the point is we should *always* fall through, https://github.com/pytorch/pytorch/issues/28386

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -165,18 +167,58 @@ class CAFFE2_API SchemaRegistrationHandleRAII final {
+  });
+}
+
+inline const KernelFunction& Dispatcher::dispatch_(const DispatchTable& dispatchTable, c10::optional<TensorTypeId> dispatchKey) const {
```
this was already inline before, just in a different file. I did a quick check on libtorch.so to confirm and actually saw a small binary size improvement from this PR, potentially because I shortened some of the call chains.

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -165,18 +167,58 @@ class CAFFE2_API SchemaRegistrationHandleRAII final {
+  });
+}
+
+inline const KernelFunction& Dispatcher::dispatch_(const DispatchTable& dispatchTable, c10::optional<TensorTypeId> dispatchKey) const {
```
Thank you, this is a lot easier to read.

One thing that is important to note is that some of the kernels you return will *never* have some of their fields set; this is not well expressed by the type information here. For example, the `catchallKernel` can *never* have the unboxed field set, because you can't possibly write a single unboxed function that will work for all types.

smessmer replied:
catchallKernels can have the unboxed field set. What you mean probably is that backend fallbacks can't have it.

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit a87be16:
* 0/2 flaky
* 1/2 broken upstream



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=a87be165aab3628a60a3e32417d1c4c10d63d835).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

ezyang commented:
I'm not sure how I feel about the new special cases for quantized tensors. Cc @zdevito

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer commented:
@ezyang they were introduced in https://github.com/pytorch/pytorch/pull/28635. I already talked to @z-a-f about them and he's creating a PR to remove them again.

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify c10 dispatcher"


Simplify the c10 dispatcher, making it more easy to understand.

Also, this moves the dispatch decision from the DispatchTable class into the Dispatcher class.
This is required because DispatchTable only knows things about one operator but the dispatch decision will (in future diffs) also need to look at backend-level fallbacks, for example for lazy.

Differential Revision: [D18018736](https://our.internmc.facebook.com/intern/diff/D18018736/)

[ghstack-poisoned]

ezyang commented on test/test_autograd.py:
```diff
@@ -3360,7 +3360,7 @@ def gradgradcheck_method_precision_override(test_name):
 def run_grad_and_gradgrad_checks(test_case, name, test_name, apply_method, output_variable,
                                  input_variables, run_gradgradcheck=True):
-    test_case.assertTrue(gradcheck(apply_method, input_variables, eps=1e-6, atol=PRECISION))
+    test_case.assertTrue(gradcheck(apply_method, input_variables, eps=1e-6, atol=PRECISION, nondet_tol=1e-10))
```
@albanD wants this change reverted
